In [1]:
#Import libraries
import numpy as np
import pandas as pd

In [2]:
#working with numpy array
arr = np.array([10, 20, 30, 40, 50, 60, 70])
print("1D Array:", arr)
print("Average Number:", arr.mean())

1D Array: [10 20 30 40 50 60 70]
Average Number: 40.0


In [3]:
#Indexing - for accessing an element on a list
print(arr[3])
print(arr[-2])
print(arr[4])
print(arr[6])

40
60
50
70


In [4]:
#Slicing - for extracting a portion of a list
print(arr[1:6])
print(arr[6:])
print(arr[:7])
print(arr[-6:-1])
print(arr[7:])
print(arr[-1:-6])

[20 30 40 50 60]
[70]
[10 20 30 40 50 60 70]
[20 30 40 50 60]
[]
[]


In [5]:
#30th April

#Creating array using np.arange 
arr3 = np.arange(1, 13)
print('Array3 Number:',arr3)

#Converting 1D array to 2D using 'Reshape'
reshape = arr3.reshape(4,3)
print('Reshaped 1D to 2D:\n', reshape)

#Converting 2D array to 1D using 'Flatten'
flatten = reshape.flatten()
print('Flatten Array:', flatten)

Array3 Number: [ 1  2  3  4  5  6  7  8  9 10 11 12]
Reshaped 1D to 2D:
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]
Flatten Array: [ 1  2  3  4  5  6  7  8  9 10 11 12]


In [6]:
#Maths Operators
arr4 = np.array([1, 2, 3, 4, 5])
arr5 = np.array([10, 20, 30, 40, 50])

#Basic Maths Operations
print('Addition:', arr4 + arr5)
print('Subtraction:', arr4 - arr5)
print('Multiplition:', arr4 * arr5)
print('Division:', arr4 / arr5)

#Scaler Operation
print('Exponential:', arr5**2)
print('Floor Div:', arr5//2)
print('Modulus:', arr5%3)

Addition: [11 22 33 44 55]
Subtraction: [ -9 -18 -27 -36 -45]
Multiplition: [ 10  40  90 160 250]
Division: [0.1 0.1 0.1 0.1 0.1]
Exponential: [ 100  400  900 1600 2500]
Floor Div: [ 5 10 15 20 25]
Modulus: [1 2 0 1 2]


In [7]:
#Dot Product using np.dot
matrix1 = np.array([
    [1, 2],
    [3, 4]
])
matrix2 = np.array([
    [5, 6],
    [7, 8]
])
dot_product = np.dot(matrix1, matrix2)
multiply = matrix1 * matrix2

print('Dot Product Result:\n', dot_product)
print('Product:\n', multiply)

Dot Product Result:
 [[19 22]
 [43 50]]
Product:
 [[ 5 12]
 [21 32]]


In [8]:
#30-day fitness tracker analysis
np.random.seed(42) #reproducibility it help generate the same set of random numbers
days = np.arange(1, 31)
steps = np.random.randint(3456, 12034, size=30)

print('Days:', days)
print('Steps:', steps)

Days: [ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30]
Steps: [10726  4316  8846  8647  9190  9721  3922  7882  9034 11778  5141  4225
 10405  5889  8767  8507  9876  4640  8011  6841  9852  6014 11305  5503
  6203  3645  6190  6461  8114  5355]


In [9]:
#Use pandas to convert the output to rows and columns using "pd.dataframe
data = pd.DataFrame({
    "Days": days,
    "Steps": steps
})

data

,Days,Steps
0,1,10726
1,2,4316
2,3,8846
3,4,8647
4,5,9190
5,6,9721
6,7,3922
7,8,7882
8,9,9034
9,10,11778


In [10]:
#Basic Stats
print('Average Step', np.mean(steps))
print('Median Step', np.median(steps))
print('Minimum Step', np.min(steps))
print('Maximum Step', np.max(steps))
print('Step Variability', np.std(steps))

Average Step 7500.2
Median Step 7946.5
Minimum Step 3645
Maximum Step 11778
Step Variability 2309.770715893679


In [11]:
#Goal analysis using np.where()
#Over 10000 Stepa
goal = np.where(steps > 10000)[0] + 1
print('Days with 10000+ Steps:', goal)



Days with 10000+ Steps: [ 1 10 13 23]


In [12]:
goal_btw_1 = np.where((steps >= 3645) & (steps <= 7956))[0] + 1
goal_btw_1

array([ 2,  7,  8, 11, 12, 14, 18, 20, 22, 24, 25, 26, 27, 28, 30])

In [13]:
goal_btw_2 = np.where((steps >= 8000) & (steps <= 11778))[0] + 1
goal_btw_2

array([ 1,  3,  4,  5,  6,  9, 10, 13, 15, 16, 17, 19, 21, 23, 29])

In [21]:
#5th May
#splitting the array to weeks using the "np.array_split()
weeks = np.array_split(steps, 4)
print(weeks)

print('Weekly Total Steps:')
week_total = [print(week.sum()) for week in weeks]

print('\nWeekly Average Steps:')
week_avg = [print(week.mean()) for week in weeks]

[array([10726,  4316,  8846,  8647,  9190,  9721,  3922,  7882],
      dtype=int32), array([ 9034, 11778,  5141,  4225, 10405,  5889,  8767,  8507],
      dtype=int32), array([ 9876,  4640,  8011,  6841,  9852,  6014, 11305], dtype=int32), array([5503, 6203, 3645, 6190, 6461, 8114, 5355], dtype=int32)]
Weekly Total Steps:
63250
63746
56539
41471

Weekly Average Steps:
7906.25
7968.25
8077.0
5924.428571428572


In [23]:
#Finding outliers(Anomaly) 
mean_step = np.mean(steps)
std_step = np.std(steps)

lower_bound = mean_step - 2*std_step
upper_bound = mean_step + 2*std_step

print(round(lower_bound, 2))
print(round(upper_bound, 2))

outliers = np.where((steps < lower_bound) | (steps > upper_bound))[0] + 1

print('Anomaly Detected:', outliers)



2880.66
12119.74
Anomaly Detected: []


The average daily steps is 7,500, meaning that on a typical day, the person walks around 7,500 steps.

The median is 7,946, a bit higher than the mean, which suggests most days are on the more active side.

The lowest day is 3,645 steps, likely a low-movement day, while the highest day reaches 11,778 marking the most active day of the month. The standard deviation is 2,309, showing that the activity varies a lot, some days are very active, others much lighter.